### ISSUE TITLE
handlebars 3.2.0 breaks existing code

### ISSUE THREAD

**Issue by Keruspe** [sunng87/handlebars-rust#348]
Presumably because of #346 but haven't confirmed yet.

This custom helper https://github.com/CleverCloud/amq-protocol/blob/011ed7a80d4448b03ca2db48b50e73c03104b98d/codegen/src/templating.rs#L283-L343 now hits this assertion with 3.2.0 while it works fine with 3.1.0 https://github.com/CleverCloud/amq-protocol/blob/011ed7a80d4448b03ca2db48b50e73c03104b98d/codegen/src/templating.rs#L307

**Comment by Keruspe** [comment]
Confirmed using bisect:

```
284c8f09ec284d877c25119f641452711ce8d85b is the first bad commit
commit 284c8f09ec284d877c25119f641452711ce8d85b
Author: Ning Sun <sunng@protonmail.com>
Date:   Fri Jun 26 17:55:50 2020 +0800

    (fix) block context for else block in each

 src/helpers/helper_each.rs | 30 +++++++++++++++++++-----------
 1 file changed, 19 insertions(+), 11 deletions(-)
```

That's not the first time that a minor semver bump breaks custom helpers

**Comment by Keruspe** [comment]
It seems that it's not actually the custom helper which is broken, but it's called with a param which is in some nested each blocks, so nested each seem to be somehow broken?

I'm trying to understand what this commit is supposed to fix and how

**Comment by sunng87** [comment]
Just realized it's a bug introduced in #346  with nested `each` block when there is empty array/object. The block context was leaked. I will create a fix soon.

**Comment by sunng87** [comment]
@Keruspe could you please help me to verify if the branch in #349 fixes the issue for you?

**Comment by Keruspe** [comment]
Sure !

I was just trying to make a minimal reproducer, will continue this way if it doesn't

**Comment by Keruspe** [comment]
Awesome, #349 seems to fix my issue!

Thanks a lot for fixing this so quickly :)

**Comment by sunng87** [comment]
Thank you! Sorry for inconvenience. This was actually a bug instead of compatibility issue. I will create a patch release asap.

**Comment by Keruspe** [comment]
@sunng87 Fwiw, in lapin I have a pretty complex template with custom helpers added on top: https://github.com/CleverCloud/lapin/blob/master/templates/channel.rs

It's pretty good at spotting this kind of stuff, feel free to ping me to do some testing (or just run ./regen-code.sh in the lapin repository) when some changes might have tricky side effects like this

**Comment by sunng87** [comment]
Nice. I have been thinking about to have CI for compatibility test against major users of handlebars-rust. I will add lapin to that system once it's ready.

**Comment by sunng87** [comment]
Fixed in #349

**Comment by sunng87** [comment]
Released in 3.2.1
